### PR TITLE
full tuning

### PR DESCRIPTION
**Pull request by SeBorgey**

### CODE REVIEW

**Comment by SeBorgey** [comment]
Посмотрите пж че mypy ругается. Нейросеть с этим не справилась.

File: autointent/modules/scoring/_transformer.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            num_labels = labels_array.shape[1]
+        else:
+            num_labels = len(set(labels))
+
+        model_name = self.model_config.model_name
+        self._tokenizer = AutoTokenizer.from_pretrained(model_name)
+        self._model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=num_labels)
+
+        def tokenize_function(examples: dict[str, Any]) -> dict[str, Any]:
+            return self._tokenizer(examples["text"], padding="max_length", truncation=True, max_length=128)
```
Comment (Samoed) [inline autointent/modules/scoring/_transformer.py]: Параметры токенизатора должны быть задаваемыми

File: autointent/modules/scoring/_transformer.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        self,
+        utterances: list[str],
+        labels: ListOfLabels,
+    ) -> None:
+        if hasattr(self, "_model"):
+            self.clear_cache()
+
+        self._validate_task(labels)
+
+        if self._multilabel:
```
Comment (Samoed) [inline autointent/modules/scoring/_transformer.py]: Значение _multilabel не меняется, поэтому ошибки mypy

File: autointent/modules/scoring/_transformer.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            training_args = TrainingArguments(
+                output_dir=tmp_dir,
+                num_train_epochs=self.num_train_epochs,
+                per_device_train_batch_size=self.batch_size,
+                learning_rate=self.learning_rate,
+                seed=self.seed,
+                save_strategy="no",
+                logging_strategy="no",
+                report_to="none",
```
Comment (Samoed) [inline autointent/modules/scoring/_transformer.py]: Мб логироване добавить?

File: autointent/modules/scoring/_transformer.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            training_args = TrainingArguments(
+                output_dir=tmp_dir,
+                num_train_epochs=self.num_train_epochs,
+                per_device_train_batch_size=self.batch_size,
+                learning_rate=self.learning_rate,
+                seed=self.seed,
+                save_strategy="no",
+                logging_strategy="no",
```
Comment (Samoed) [inline autointent/modules/scoring/_transformer.py]: И я тут бы none тоже не делал

File: autointent/modules/scoring/_transformer.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+        if self._multilabel:
+            labels_array = np.array(labels) if not isinstance(labels, np.ndarray) else labels
+            num_labels = labels_array.shape[1]
+        else:
+            num_labels = len(set(labels))
+
+        model_name = self.model_config.model_name
+        self._tokenizer = AutoTokenizer.from_pretrained(model_name)
+        self._model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=num_labels)
```
Comment (Samoed) [inline autointent/modules/scoring/_transformer.py]: Модель на device не перенес (можно вроде в тренер передавать)

File: tests/modules/scoring/test_transformer.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        assert metadata is None
+    else:
+        pytest.skip("predict_with_metadata not implemented in TransformerScorer")
+
+    scorer.clear_cache()
+    assert not hasattr(scorer, "_model") or scorer._model is None
+    assert not hasattr(scorer, "_tokenizer") or scorer._tokenizer is None
+
+    with pytest.raises(RuntimeError):
+        scorer.predict(test_data)
```
Comment (Samoed) [inline tests/modules/scoring/test_transformer.py]: Ты знаешь все параметры уже и не надо доп проверок

```suggestion
```

**Review by Samoed — CHANGES_REQUESTED** [review]

File: autointent/modules/scoring/_transformer.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        utterances: list[str],
+        labels: ListOfLabels,
+    ) -> None:
+        if hasattr(self, "_model"):
+            self.clear_cache()
+
+        self._validate_task(labels)
+
+        if self._multilabel:
+            labels_array = np.array(labels) if not isinstance(labels, np.ndarray) else labels
```
Comment (voorhs) [inline autointent/modules/scoring/_transformer.py]: labels это всегда список строк, никогда не бывает нампай массив, поэтому эта проверка не имеет смысла, возможны ошибки mypy

File: autointent/modules/scoring/_transformer.py (hunk 1 above)
Comment (voorhs) [reply autointent/modules/scoring/_transformer.py]: надо бы завести что-то вроде `TokenizerConfig` и подавать его как параметр конструктора

File: autointent/modules/scoring/_transformer.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+    @classmethod
+    def from_context(
+        cls,
+        context: Context,
+        model_config: EmbedderConfig | str | None = None,
+    ) -> "TransformerScorer":
+        if model_config is None:
+            model_config = context.resolve_embedder()
+        return cls(model_config=model_config)
```
Comment (voorhs) [inline autointent/modules/scoring/_transformer.py]: здесь должны быть все те же параметры, что и в конструкторе

вместе с теми же дефолтными значениями

с помощью этой фабрики инициализируется модуль во время процесса подбора гиперпарметров

File: autointent/modules/scoring/_transformer.py (hunk 4 above)
Comment (voorhs) [reply autointent/modules/scoring/_transformer.py]: @Samoed как думаешь, возможно как-то логировать кривые обучения в глобальный wandb проект который используется во время всего процесса оптимизации?

File: autointent/modules/scoring/_transformer.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+            trainer = Trainer(
+                model=self._model,
+                args=training_args,
+                train_dataset=tokenized_dataset,
+                tokenizer=self._tokenizer,
+                data_collator=DataCollatorWithPadding(tokenizer=self._tokenizer),
+            )
+
+            trainer.train()
```
Comment (voorhs) [inline autointent/modules/scoring/_transformer.py]: насколько я понимаю нужно написать `Trainer(model=self._model.to(device), ...)`, чтобы обучение производилось на гпу

возможно не прав, скажи потом если у тебя оно само на гпу перекидывалось

Comment (voorhs) [reply autointent/modules/scoring/_transformer.py]: желательно просто этот параметр инферить из `EmbedderConfig`

File: autointent/modules/scoring/_transformer.py
Comment (voorhs) [inline autointent/modules/scoring/_transformer.py:1]: лучше назвать `BERTScorer`

File: tests/modules/scoring/test_transformer.py (hunk 6 above)
Comment (voorhs) [reply tests/modules/scoring/test_transformer.py]: а по-моему ок, пусть будет

File: autointent/modules/scoring/_transformer.py (hunk 7 above)
Comment (voorhs) [reply autointent/modules/scoring/_transformer.py]: да, это источник ошибки
```
autointent/modules/scoring/_transformer.py:70: error: Subclass of "list[int]" and "ndarray[Any, Any]" cannot exist: would have incompatible method signatures  [unreachable]
autointent/modules/scoring/_transformer.py:70: error: Subclass of "list[list[int]]" and "ndarray[Any, Any]" cannot exist: would have incompatible method signatures  [unreachable]
```

File: autointent/modules/scoring/_transformer.py (hunk 1 above)
Comment (voorhs) [reply autointent/modules/scoring/_transformer.py]: а еще здесь надо в 80 строчке написать `# type: ignore[no-any-return]`, потому что возникает вот такая ошибка:
```
autointent/modules/scoring/_transformer.py:80: error: Returning Any from function declared to return "dict[str, Any]"  [no-any-return]
```

File: autointent/modules/scoring/_transformer.py (hunk 3 above)
Comment (voorhs) [reply autointent/modules/scoring/_transformer.py]: `report_to="wandb"` надо попробовать

File: autointent/modules/scoring/_transformer.py (hunk 9 above)
Comment (voorhs) [reply autointent/modules/scoring/_transformer.py]: https://huggingface.co/docs/transformers/v4.49.0/en/main_classes/trainer#transformers.Seq2SeqTrainingArguments.use_cpu
```
use_cpu=(self.model_config.device == "cpu")
```

File: tests/modules/scoring/test_transformer.py (hunk 6 above)
Comment (voorhs) [reply tests/modules/scoring/test_transformer.py]: убрать `pytest.skip`

дописать комментарии что тестируется

File: autointent/modules/scoring/_transformer.py
Comment (Samoed) [reply autointent/modules/scoring/_transformer.py:1]: Мб поменять на EncoderScorer, тк не только берты?

File: autointent/modules/scoring/_bert.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+        learning_rate: float = 5e-5,
+        seed: int = 0,
+        tokenizer_config: TokenizerConfig | None = None,
+    ) -> None:
+        self.model_config = EmbedderConfig.from_search_config(model_config)
+        self.num_train_epochs = num_train_epochs
+        self.batch_size = batch_size
+        self.learning_rate = learning_rate
+        self.seed = seed
+        self.tokenizer_config = tokenizer_config or TokenizerConfig()
```
Comment (Samoed) [inline autointent/modules/scoring/_bert.py]: Я бы оставил его None, если он не передается, тк лучше его не менять специально

File: autointent/modules/scoring/_bert.py [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+        self._multilabel = False
```
Comment (Samoed) [inline autointent/modules/scoring/_bert.py]: Мб передвинуть к supports?

File: autointent/modules/scoring/_bert.py [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+            labels_array = np.array(labels)
+            num_labels = labels_array.shape[1]
+        else:
+            num_labels = len(set(labels))
+
+        model_name = self.model_config.model_name
+        self._tokenizer = AutoTokenizer.from_pretrained(model_name)
+        self._model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=num_labels)
+
+        use_cpu = hasattr(self.model_config, "device") and self.model_config.device == "cpu"
```
Comment (Samoed) [inline autointent/modules/scoring/_bert.py]: `model_config` всегда будет иметь `device`

File: autointent/modules/scoring/_bert.py [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+            training_args = TrainingArguments(
+                output_dir=tmp_dir,
+                num_train_epochs=self.num_train_epochs,
+                per_device_train_batch_size=self.batch_size,
+                learning_rate=self.learning_rate,
+                seed=self.seed,
+                save_strategy="no",
+                logging_strategy="steps",
+                logging_steps=10,
+                report_to="wandb",
```
Comment (Samoed) [inline autointent/modules/scoring/_bert.py]: Тут надо не только wandb, а то что задано в logging_config. Его можно получить из контекста при инициализации

File: autointent/modules/scoring/_bert.py [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+
+        use_cpu = hasattr(self.model_config, "device") and self.model_config.device == "cpu"
+
+        def tokenize_function(examples: dict[str, Any]) -> dict[str, Any]:
+            return self._tokenizer(  # type: ignore[no-any-return]
+                examples["text"],
+                padding=self.tokenizer_config.padding,
+                truncation=self.tokenizer_config.truncation,
+                max_length=self.tokenizer_config.max_length,
+            )
```
Comment (Samoed) [inline autointent/modules/scoring/_bert.py]:
```suggestion
            return self._tokenizer(  # type: ignore[no-any-return]
                examples["text"],
                padding=self.tokenizer_config.padding,
                truncation=self.tokenizer_config.truncation,
                max_length=self.tokenizer_config.max_length,
                return_tensors="pt"
            )
```

File: autointent/modules/scoring/_bert.py [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+
+        self._model.eval()
+
+    def predict(self, utterances: list[str]) -> npt.NDArray[Any]:
+        if not hasattr(self, "_model") or not hasattr(self, "_tokenizer"):
+            msg = "Model is not trained. Call fit() first."
+            raise RuntimeError(msg)
+
+        inputs = self._tokenizer(
+            utterances, padding=True, truncation=True, max_length=self.tokenizer_config.max_length, return_tensors="pt"
```
Comment (Samoed) [inline autointent/modules/scoring/_bert.py]:
```suggestion
            utterances, padding=self.tokenizer_config.padding, truncation=self.tokenizer_config.truncation, max_length=self.tokenizer_config.max_length, return_tensors="pt"
```

File: autointent/modules/scoring/_transformer.py
Comment (voorhs) [reply autointent/modules/scoring/_transformer.py:1]: ну по-моему BERTScorer самое точное название поскольку тут предполагается использовать BERT-like модели

Comment (Samoed) [reply autointent/modules/scoring/_transformer.py:1]: Тут просто AutoModelForSequenceClassification и там может быть по сути что угодно

File: autointent/modules/scoring/_bert.py (hunk 11 above)
Comment (voorhs) [reply autointent/modules/scoring/_bert.py]: @SeBorgey не очень понимаю зачем изначально ты это сделал

этот атрибут инициализируется при вызове `self._validaite_task(labels)`

File: autointent/modules/scoring/_bert.py [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+            tokenizer_config=tokenizer_config,
+        )
+
+    def get_embedder_config(self) -> dict[str, Any]:
+        return self.model_config.model_dump()
+
+    def _validate_task(self, labels: ListOfLabels) -> None:
+        """Validate the task and set _multilabel flag."""
+        super()._validate_task(labels)
+        self._multilabel = isinstance(labels[0], list)
```
Comment (voorhs) [inline autointent/modules/scoring/_bert.py]: `self._multilabel` и так устанавливается по такому правилу внутри `super()._validate_task(labels)`

код вот тут: https://github.com/deeppavlov/AutoIntent/blob/feat/transformer-scorer/autointent/modules/base/_base.py#L234

File: autointent/modules/scoring/_bert.py (hunk 12 above)
Comment (voorhs) [reply autointent/modules/scoring/_bert.py]: да, скорее всего из-за этой mypy выдаст ошибку unreachable, проверку на наличие атрибута надо убрать

File: autointent/modules/scoring/_bert.py [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+        if hasattr(self, "_model"):
+            self.clear_cache()
+
+        self._validate_task(labels)
+
+        if self._multilabel:
+            labels_array = np.array(labels)
+            num_labels = labels_array.shape[1]
+        else:
+            num_labels = len(set(labels))
```
Comment (voorhs) [inline autointent/modules/scoring/_bert.py]: после работы `self._validate_task(label)` сохраняется атрибут `self._n_classes`, тебе не надо самому это считать

File: autointent/modules/scoring/_bert.py (hunk 13 above)
Comment (voorhs) [reply autointent/modules/scoring/_bert.py]: @SeBorgey добавь в конструктор класса аргумент `report_to: REPORTERS_NAMES | None = None`

а в метод `from_context` добавь следующий код:
```python
from autointent._callbacks import REPORTERS_NAMES

report_to = context.logging_config.report_to
return cls(..., report_to=report_to)
```

File: autointent/modules/scoring/_bert.py [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+            msg = "Model is not trained. Call fit() first."
+            raise RuntimeError(msg)
+
+        inputs = self._tokenizer(
+            utterances, padding=True, truncation=True, max_length=self.tokenizer_config.max_length, return_tensors="pt"
+        )
+
+        with torch.no_grad():
+            outputs = self._model(**inputs)
+            logits = outputs.logits
```
Comment (voorhs) [inline autointent/modules/scoring/_bert.py]: на инференсе тоже не стоит сразу все подавать разом, лучше разбить по батчам

Comment (Samoed) [reply autointent/modules/scoring/_bert.py]: Можно использовать pipeline, чтобы меньше писать

**Review by voorhs — COMMENTED** [review]
а ты пробовал прогонять на гпу? я просто в predict не вижу переноса батчей на девайс. как будто на cpu код всегда будет работать а с гпу нет

File: autointent/modules/scoring/_bert.py [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+        labels: ListOfLabels,
+    ) -> None:
+        if hasattr(self, "_model"):
+            self.clear_cache()
+
+        self._validate_task(labels)
+
+        model_name = self.model_config.model_name
+        self._tokenizer = AutoTokenizer.from_pretrained(model_name)
+        self._model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=self._n_classes)
```
Comment (voorhs) [inline autointent/modules/scoring/_bert.py:87]: label2id

File: autointent/modules/scoring/_bert.py [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+            batch = utterances[i : i + self.batch_size]
+            inputs = self._tokenizer(batch, return_tensors="pt", **self.model_config.tokenizer_config.model_dump())
+            with torch.no_grad():
+                outputs = self._model(**inputs)
+                logits = outputs.logits
+            if self._multilabel:
+                batch_predictions = torch.sigmoid(logits).numpy()
+            else:
+                batch_predictions = torch.softmax(logits, dim=1).numpy()
+            all_predictions.append(batch_predictions)
```
Comment (voorhs) [inline autointent/modules/scoring/_bert.py:141]: https://huggingface.co/docs/transformers/main/en/main_classes/pipelines#pipelines
